File: mobile/modules/bluetooth-sdk/android/src/main/java/com/mentra/bluetoothsdk/sgcs/MentraLivePairingAdvertisement.kt
package com.mentra.bluetoothsdk.sgcs

internal data class MentraLivePairingAdvertisement(
    val pairingMode: Boolean,
    val pairingCode: String,
)

internal object MentraLivePairingAdvertisementParser {
    private const val PAIRING_FLAG_OFFSET = 5
    private const val PAIRING_DISCOVERABLE: Byte = 0x01
    private const val PROTOCOL_VERSION_OFFSET = PAIRING_FLAG_OFFSET + 1
    private const val CAPABILITY_OFFSET = PROTOCOL_VERSION_OFFSET + 1
    private const val CODE_LOW_OFFSET = CAPABILITY_OFFSET + 1
    private const val CODE_HIGH_OFFSET = CODE_LOW_OFFSET + 1
    private const val MAGIC_FIRST_OFFSET = CODE_HIGH_OFFSET + 1
    private const val MAGIC_SECOND_OFFSET = MAGIC_FIRST_OFFSET + 1
    private const val PROTOCOL_VERSION_MIN = 2
    private const val PROTOCOL_VERSION_MAX = 15
    private const val SECURE_PAIRING_CAPABILITY = 0x01
    private const val MAGIC_FIRST = 0x4D // M
    private const val MAGIC_SECOND = 0x50 // P

    /**
     * Parses company-id-stripped 0xB822 manufacturer data.
     *
     * Legacy firmware stores an XOR'd Classic MAC where the original pairing implementation
     * expected version and capability bytes. Requiring the `MP` marker makes the formats
     * unambiguous instead of probabilistically classifying MAC bytes as a secure trailer.
     */
    fun parse(manufacturerData: ByteArray?): MentraLivePairingAdvertisement? {
        if (manufacturerData == null || manufacturerData.size <= MAGIC_SECOND_OFFSET) {
            return null
        }

        val version = manufacturerData[PROTOCOL_VERSION_OFFSET].toInt() and 0xFF
        val capability = manufacturerData[CAPABILITY_OFFSET].toInt() and 0xFF
        val magicFirst = manufacturerData[MAGIC_FIRST_OFFSET].toInt() and 0xFF
        val magicSecond = manufacturerData[MAGIC_SECOND_OFFSET].toInt() and 0xFF
        if (version !in PROTOCOL_VERSION_MIN..PROTOCOL_VERSION_MAX ||
            (capability and SECURE_PAIRING_CAPABILITY) == 0 ||
            magicFirst != MAGIC_FIRST ||
            magicSecond != MAGIC_SECOND
        ) {
            return null
        }

        val codeLow = manufacturerData[CODE_LOW_OFFSET].toInt() and 0xFF
        val codeHigh = manufacturerData[CODE_HIGH_OFFSET].toInt() and 0xFF
        return MentraLivePairingAdvertisement(
            pairingMode = manufacturerData[PAIRING_FLAG_OFFSET] == PAIRING_DISCOVERABLE,
            pairingCode = String.format("%02X%02X", codeHigh, codeLow),
        )
    }
}

File: mobile/modules/bluetooth-sdk/ios/Source/sgcs/MentraLivePairingAdvertisement.swift
import Foundation

struct MentraLivePairingAdvertisement: Equatable {
    let pairingMode: Bool
    let pairingCode: String

    private static let manufacturerId: UInt16 = 0xB822
    private static let companyIdLength = 2
    private static let pairingFlagOffset = 5
    private static let pairingDiscoverable: UInt8 = 0x01
    private static let protocolVersionOffset = pairingFlagOffset + 1
    private static let capabilityOffset = protocolVersionOffset + 1
    private static let codeLowOffset = capabilityOffset + 1
    private static let codeHighOffset = codeLowOffset + 1
    private static let magicFirstOffset = codeHighOffset + 1
    private static let magicSecondOffset = magicFirstOffset + 1
    private static let protocolVersionRange = 2 ... 15
    private static let securePairingCapability = 0x01
    private static let magicFirst: UInt8 = 0x4D // M
    private static let magicSecond: UInt8 = 0x50 // P

    /// Parses CoreBluetooth manufacturer data, including its two-byte company-id prefix.
    ///
    /// Legacy firmware stores an XOR'd Classic MAC where the original pairing implementation
    /// expected version and capability bytes. Requiring the `MP` marker makes the formats
    /// unambiguous instead of probabilistically classifying MAC bytes as a secure trailer.
    static func parse(coreBluetoothManufacturerData data: Data?) -> MentraLivePairingAdvertisement? {
        guard let data, data.count > companyIdLength + magicSecondOffset else {
            return nil
        }

        let companyId = UInt16(data[0]) | (UInt16(data[1]) << 8)
        guard companyId == manufacturerId else {
            return nil
        }

        let payloadBase = companyIdLength
        let version = Int(data[payloadBase + protocolVersionOffset])
        let capability = Int(data[payloadBase + capabilityOffset])
        guard protocolVersionRange.contains(version),
              capability & securePairingCapability != 0,
              data[payloadBase + magicFirstOffset] == magicFirst,
              data[payloadBase + magicSecondOffset] == magicSecond
        else {
            return nil
        }

        let codeLow = Int(data[payloadBase + codeLowOffset])
        let codeHigh = Int(data[payloadBase + codeHighOffset])
        return MentraLivePairingAdvertisement(
            pairingMode: data[payloadBase + pairingFlagOffset] == pairingDiscoverable,
            pairingCode: String(format: "%02X%02X", codeHigh, codeLow)
        )
    }
}

enum MentraLiveConnectionAttemptPolicy {
    static func shouldAcceptDidConnect(
        pairingYieldActive: Bool,
        matchesActiveAttempt: Bool
    ) -> Bool {
        !pairingYieldActive && matchesActiveAttempt
    }
}
